<!DOCTYPE html>
<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">

  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <title>jianhong/16S_pipeline</title>
  <meta name="description" content="This pipeline analyses data for microbiome 16S data.">
  <meta name="author" content="Phil Ewels">
  <meta name="color-scheme" content="light dark">
  <link rel="shortcut icon" href="assets/nf-core-logo-square.png" type="image/png">
  <link rel="alternate" type="application/rss+xml" title="nf-core: Events" href="./events/rss">
  <!-- FontAwesome -->
  <script src="https://kit.fontawesome.com/471b59d3f8.js"></script>
  <script src="assets/jquery.min.js"></script>
  <script src="assets/bootstrap.bundle.min.js"></script>
  <script src="assets/highlight.min.js"></script>
  <script src="assets/groovy.min.js"></script>
  <!-- Page-specific CSS and JS -->
      <script src="assets/moment.js"></script>
    <script src="assets/Chart.min.js"></script>
    <link href="assets/Chart.min.css" rel="stylesheet">
    <script src="assets/hammer.min.js"></script>
    <script src="assets/chartjs-plugin-zoom.min.js"></script>
    <!-- NO CURRENT VERSION ON NPM --> <script src="assets/canvas2svg.js"></script><!-- NO CURRENT VERSION ON NPM -->
    <script src="assets/FileSaver.min.js"></script>
    <script src="assets/jquery.tablesorter.min.js"></script>

  <!-- Custom nf-core CSS and JS -->
  <link href="assets/nf-core-light.css" rel="stylesheet" id="theme-stylesheet">
  <script src="assets/nf-core.js"></script>
  <script>
    window.dataLayer = window.dataLayer || [];

    function gtag() {
      dataLayer.push(arguments);
    }
      </script>
<style type="text/css">/* Chart.js */
@keyframes chartjs-render-animation{from{opacity:.99}to{opacity:1}}.chartjs-render-monitor{animation:chartjs-render-animation 1ms}.chartjs-size-monitor,.chartjs-size-monitor-expand,.chartjs-size-monitor-shrink{position:absolute;direction:ltr;left:0;top:0;right:0;bottom:0;overflow:hidden;pointer-events:none;visibility:hidden;z-index:-1}.chartjs-size-monitor-expand>div{position:absolute;width:1000000px;height:1000000px;left:0;top:0}.chartjs-size-monitor-shrink>div{position:absolute;width:200%;height:200%;left:0;top:0}</style>

</head>

<body tabindex="0" data-new-gr-c-s-check-loaded="14.1047.0" data-gr-ext-installed="">
  <nav class="navbar fixed-top navbar-expand-md navbar-light bg-light shadow-sm site-nav d-print-none">
    <div class="container-fluid">
      <a class="navbar-brand d-md-none" href="./">
        <img height="25px" src="assets/nf-core-logo.svg" class="hide-dark">
        <img height="25px" src="assets/nf-core-logo-darkbg.svg" class="hide-light">
      </a>
      <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation">
        <span class="navbar-toggler-icon"></span>
      </button>
      <div class="collapse navbar-collapse justify-content-md-center" id="navbarCollapse">
        <ul class="navbar-nav">
          <li class="nav-item p-1">
            <a class="nav-link" href="index.html">Home</a>
          </li>
          <li class="nav-item p-1">
            <a class="nav-link" href="usage.html">Usage docs</a>
          </li>
          <li class="nav-item p-1">
            <a class="nav-link" href="parameters.html">Parameters</a>
          </li>
          <li class="nav-item p-1">
            <a class="nav-link" href="output.html">Output docs</a>
          </li>
          <li class="nav-item p-1">
            <a class="nav-link" href="multiqc.html">Multiqc</a>
          </li>
          <li class="nav-item p-1">
            <a class="nav-link" href="krona.html">Krona</a>
          </li>
        </ul>
        <hr class="d-md-none">
      </div>
    </div>
  </nav>


  <div class="mainpage">
    <div class="mainpage-heading triangle-down">
      <div class="container">
        <h1 class="display-2">jianhong/<br class="d-sm-none">16S_pipeline</h1>
        <p class="lead">This pipeline analyses data for 16S microbiome.</p>
      </div>
    </div>

  <div class="container-fluid mainpage-subheader-heading chevron-down">
    <div class="container text-center">
      <p>
        <a href="https://github.com/jianhong/16S_pipeline" class="btn btn-success btn-lg d-print-none">
          <i class="fab fa-github" aria-hidden="true"></i> https://github.com/jianhong/16S_pipeline
        </a>
      </p>
    </div>
  </div>

<div class="container-xxl main-content">

  <ul class="nav nav-fill nfcore-subnav justify-content-around d-print-none">
    <li class="nav-item">
      <a class="nav-link active" href="index.html"><i class="fas fa-sign-in me-1" aria-hidden="true"></i> Introduction</a>
    </li>
        <li class="nav-item">
      <a class="nav-link" href="usage.html"><i class="far fa-book me-1" aria-hidden="true"></i> Usage docs</a>
    </li>
          <li class="nav-item">
        <a class="nav-link" href="parameters.html"><i class="far fa-book me-1" aria-hidden="true"></i> Parameter docs</a>
      </li>
        <li class="nav-item">
      <a class="nav-link" href="output.html"><i class="far fa-book me-1" aria-hidden="true"></i> Output docs</a>
    </li>
    <li class="nav-item">
    <a class="nav-link" href="multiqc.html"><i class="far fa-book me-1" aria-hidden="true"></i> MultiQC </a>
    </li>
    <li class="nav-item">
    <a class="nav-link" href="krona.html"><i class="far fa-book me-1" aria-hidden="true"></i> Krona </a>
    </li>

  </ul>

  <div class="row flex-wrap-reverse flex-lg-wrap ms-lg-5"><div class="col-12 col-lg-9"><div class="rendered-markdown pipeline-page-content"><h1 id="introduction" class="">Introduction<a href="#introduction" class="header-link scroll_to_link"><span class="fas fa-link"></span></a></h1>
<p><strong>16S_pipeline</strong> is a bioinformatics best-practice analysis pipeline for 16S rRNA gene sequencing.</p>
<p>The pipeline is built using <a href="https://www.nextflow.io">Nextflow</a>,
 a workflow tool to run tasks across multiple compute infrastructures in a very
  portable manner. It uses Docker/Singularity containers making installation
  trivial and results highly reproducible.
  The <a href="https://www.nextflow.io/docs/latest/dsl2.html">Nextflow DSL2</a>
  implementation of this pipeline uses one container per process which makes it
  much easier to maintain and update software dependencies. </p>


<h2 id="pipeline-summary" class="">Pipeline summary<a href="#pipeline-summary" class="header-link scroll_to_link"><span class="fas fa-link"></span></a></h2>

<ol start="0">
<li>Prepare fastq files (<a href="https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html"><code>bcl2fastq</code></a>)</li>
<li>Read QC (<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"><code>FastQC</code></a>)</li>
<li>Remove primers (<a href="http://www.usadellab.org/cms/?page=trimmomatic"><code>trimmomatic</code></a>)</li>
<li>Sync barcodes (<a href="https://gist.github.com/588841/"><code>fastq_pair_filter.py</code></a>)</li>
<li>Demultiplex (<a href="https://docs.qiime2.org/2021.11/plugins/available/demux/"><code>qiime2::demux</code></a>)</li>
<li>Filter reads (<a href="http://benjjneb.github.io/dada2/"><code>DATA2</code></a>)</li>
<li>Run dada2 (<a href="http://benjjneb.github.io/dada2/"><code>DATA2</code></a>)</li>
<li>Visualization (<a href="https://github.com/marbl/Krona/wiki"><code>Krona</code></a>)</li>
<li>Present QC for raw reads (<a href="http://multiqc.info/"><code>MultiQC</code></a>)</li>
</ol>
<h2 id="quick-start" class="">Quick Start<a href="index.html#quick-start" class="header-link scroll_to_link"><span class="fas fa-link" aria-hidden="true"></span></a></h2>
<ol>
<li>
<p>Install <a href="https://www.nextflow.io/docs/latest/getstarted.html#installation"><code>Nextflow</code></a> (<code>&gt;=21.10.3</code>)</p>
</li>
<li>
<p>Install any of <a href="https://docs.docker.com/engine/installation/"><code>Docker</code></a>, <a href="https://www.sylabs.io/guides/3.0/user-guide/"><code>Singularity</code></a>, <a href="https://podman.io/"><code>Podman</code></a>, <a href="https://nersc.gitlab.io/development/shifter/how-to-use/"><code>Shifter</code></a> or <a href="https://hpc.github.io/charliecloud/"><code>Charliecloud</code></a> for full pipeline reproducibility <em>(please only use <a href="https://conda.io/miniconda.html"><code>Conda</code></a> as a last resort; see <a href="https://nf-co.re/usage/configuration#basic-configuration-profiles">docs</a>)</em></p>
</li>
<li>
<p>Download the pipeline and test it on a minimal dataset with a single command:</p>
<pre><code class="language-console hljs language-shell"> nextflow run jianhong/16S_pipeline -profile test,YOURPROFILE</code></pre>
<p>Note that some form of configuration will be needed so that Nextflow knows how to fetch the required software. This is usually done in the form of a config profile (<code>YOURPROFILE</code> in the example command above). You can chain multiple config profiles in a comma-separated string.</p>
<blockquote>
<ul>
<li>The pipeline comes with config profiles called <code>docker</code>, <code>singularity</code>, <code>podman</code>, <code>shifter</code>, <code>charliecloud</code> and <code>conda</code> which instruct the pipeline to use the named tool for software management. For example, <code>-profile test,docker</code>.</li>
<li>Please check <a href="https://github.com/nf-core/configs#documentation">nf-core/configs</a> to see if a custom config file to run nf-core pipelines already exists for your Institute. If so, you can simply use <code>-profile &lt;institute&gt;</code> in your command. This will enable either <code>docker</code> or <code>singularity</code> and set the appropriate execution settings for your local compute environment.</li>
<li>If you are using <code>singularity</code> and are persistently observing issues downloading Singularity images directly due to timeout or network issues, then you can use the <code><span class="text-nowrap">--singularity_pull_docker_container</span></code> parameter to pull and convert the Docker image instead. Alternatively, you can use the <a href="https://nf-co.re/tools/#downloading-pipelines-for-offline-use"><code>nf-core download</code></a> command to download images first, before running the pipeline. Setting the <a href="https://www.nextflow.io/docs/latest/singularity.html?#singularity-docker-hub"><code>NXF_SINGULARITY_CACHEDIR</code> or <code>singularity.cacheDir</code></a> Nextflow options enables you to store and re-use the images from a central location for future pipeline runs.</li>
<li>If you are using <code>conda</code>, it is highly recommended to use the <a href="https://www.nextflow.io/docs/latest/conda.html"><code>NXF_CONDA_CACHEDIR</code> or <code>conda.cacheDir</code></a> settings to store the environments in a central location for future pipeline runs.</li>
</ul>
</blockquote>
</li>
<li>
<p>Start running your own analysis!</p>
<pre><code class="language-console hljs language-shell"> nextflow run jianhong/16S_pipeline -profile &lt;docker/singularity/podman/shifter/charliecloud/conda/institute&gt; --input '[path to raw reads files]' --barcodes '[path to barcodes tsv file]' --metadata '[path to metadata csv file]'</code></pre>
<p>Run it on cluster.</p>
<p>First prepare a profile config file named as <a href="usage.html">profile.config</a>.</p>
<pre><code class="language-console hljs language-shell"> // submit by slurm
 process.executor = "slurm"
 process.clusterOptions = "-J microbiome"
 params {
     // Input data
     input  = 'path/to/your/fastqfiles'
     skip_bcl2fastq = true
     barcodes = 'path/to/your/barcodes.tsv'
     metadata = 'path/to/your/metadata.csv'

     // Genome references
     silva_nr99 = 'https://zenodo.org/record/4587955/files/silva_nr99_v138.1_train_set.fa.gz?download=1'
     silva_tax = 'https://zenodo.org/record/4587955/files/silva_species_assignment_v138.1.fa.gz?download=1'
 }</code></pre>
<p>Then run:</p>
<pre><code class="language-console hljs language-shell"> nextflow run jianhong/16S_pipeline -profile &lt;docker/singularity/podman/shifter/charliecloud/conda/institute&gt; -c profile.config</code></pre>
</li>
</ol>
<h2 id="create-conda-container-for-nextflow" class="">Create <code>conda</code> container for <code>nextflow</code><a href="index.html#create-conda-container-for-nextflow" class="header-link scroll_to_link"><span class="fas fa-link" aria-hidden="true"></span></a></h2>
<ol>
<li>install <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html"><code>conda</code></a></li>
</ol>
<pre><code class="language-console hljs language-shell">wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda.sh
bash Miniconda3-latest-Linux-x86_64.sh</code></pre>
<ol start="2">
<li>Create <a href="https://www.nextflow.io/"><code>nextflow</code></a> environment.</li>
</ol>
<pre><code class="language-console hljs language-shell">conda create -y --name microbiome bioconda::nextflow=21.10.6</code></pre>
<ol start="3">
<li>Create profile config file named as <a href="usage.html">profile.config</a>.</li>
</ol>
<pre><code class="language-console hljs language-shell">// submit by slurm
process.executor = "slurm"
process.clusterOptions = "-J JL21"
params {
    // Input data
    input  = 'Spinach_done' // replace Spinach_done by your own file
    barcodes = '16S_pipeline_JL21/0_mapping/barcodes.tsv'
    metadata = '16S_pipeline_JL21/0_mapping/metadata.csv'

    // report email
    email = 'your@email.addr'
}</code></pre>
<ol start="4">
<li>Activate <a href="https://www.nextflow.io/"><code>nextflow</code></a> environment and Run the pipeline.</li>
</ol>
<pre><code class="language-console hljs language-shell">conda activate microbiome
module load bcl2fastq/2.20
nextflow run jianhong/16S_pipeline -r main -profile conda -c profile.config</code></pre>
<h2 id="documentation" class="">Documentation<a href="index.html#documentation" class="header-link scroll_to_link"><span class="fas fa-link" aria-hidden="true"></span></a></h2>
<p>The 16S_pipeline pipeline comes with documentation about the pipeline <a href="usage.html">usage</a>, and <a href="output.html">output</a>.</p>
<h2 id="credits" class="">Credits<a href="index.html#credits" class="header-link scroll_to_link"><span class="fas fa-link" aria-hidden="true"></span></a></h2>
<p>16S_pipeline was originally written by Jianhong Ou, and Jeff Letourneau.</p>
<p>We thank the following people for their extensive assistance in the development of this pipeline:</p>
<!-- TODO nf-core: If applicable, make list of people who have also contributed -->
<h2 id="contributions-and-support" class="">Contributions and Support<a href="index.html#contributions-and-support" class="header-link scroll_to_link"><span class="fas fa-link" aria-hidden="true"></span></a></h2>
<p>If you would like to contribute to this pipeline, please see the <a href="https://github.com/jianhong/16S_pipeline/blob/main/.github/CONTRIBUTING.html">contributing guidelines</a>.</p>
<h2 id="citations" class="">Citations<a href="index.html#citations" class="header-link scroll_to_link"><span class="fas fa-link" aria-hidden="true"></span></a></h2>
<!-- TODO nf-core: Add citation for pipeline after first release. Uncomment lines below and update Zenodo doi and badge at the top of this file. -->
<!-- If you use  16S_pipeline for your analysis, please cite it using the following doi: [10.5281/zenodo.XXXXXX](https://doi.org/10.5281/zenodo.XXXXXX) -->
<!-- TODO nf-core: Add bibliography of tools and data used in your pipeline -->
<p>An extensive list of references for the tools used by the pipeline can be found in the <a href="https://github.com/jianhong/16S_pipeline/blob/main/CITATIONS.md"><code>CITATIONS.md</code></a> file.</p>
<p>You can cite the <code>nf-core</code> publication as follows:</p>
<blockquote>
<p><strong>The nf-core framework for community-curated bioinformatics pipelines.</strong></p>
<p>Philip Ewels, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso &amp; Sven Nahnsen.</p>
<p><em>Nat Biotechnol.</em> 2020 Feb 13. doi: <a href="https://dx.doi.org/10.1038/s41587-020-0439-x">10.1038/s41587-020-0439-x</a>.</p>
</blockquote></div></div><div class="col-12 col-lg-3 ps-2"><div class="side-sub-subnav sticky-top">
<div class="pipeline-sidebar">
  <div class="row border-bottom pb-2">
    <div class="col-12">
      <h6><i class="fas fa-terminal fa-xs" aria-hidden="true"></i> command</h6>
      <div class="input-group input-group-sm pipeline-run-cmd">
        <input type="text" class="form-control input-sm code" id="pipeline-run-cmd-text" data-autoselect="" value="nextflow run jianhong/16S_pipeline -r main -profile test" aria-label="Copy run command" readonly="">
        <button class="btn btn-outline-secondary copy-txt" data-bs-target="pipeline-run-cmd-text" data-bs-toggle="tooltip" data-bs-placement="left" title="Copy to clipboard" type="button"><i class="fas fa-clipboard px-1" aria-hidden="true"></i></button>
      </div>
    </div>
  </div>

  <div class="row border-bottom">
  <div>
    <h6>get in touch</h6>
    <p><a class="btn btn-sm btn-outline-secondary" href="https://github.com/jianhong/16S_pipeline/issues"><i class="fab fa-github me-1" aria-hidden="true"></i> Open an issue on GitHub</a></p>
  </div>
</div>

</div></div></div></div></div>

<footer class="subfooter"><div class="container"><p class="mb-0"><i class="fab fa-github" aria-hidden="true"></i> Read this page on GitHub: <code><a href="https://github.com/jianhong/16S_pipeline/blob/main/README.md">https://github.com/jianhong/16S_pipeline/blob/main/README.md</a></code></p></div></footer>


<script type="text/javascript">
  $(function() {
    // Plot hover vertical line
    // https://stackoverflow.com/a/45172506/713980
    Chart.defaults.LineWithLine = Chart.defaults.line;
    Chart.controllers.LineWithLine = Chart.controllers.line.extend({
      draw: function(ease) {
        Chart.controllers.line.prototype.draw.call(this, ease);

        if (this.chart.tooltip._active && this.chart.tooltip._active.length) {
          var activePoint = this.chart.tooltip._active[0],
            ctx = this.chart.ctx,
            x = activePoint.tooltipPosition().x,
            topY = this.chart.scales['y-axis-0'].top,
            bottomY = this.chart.scales['y-axis-0'].bottom;

          // draw line
          ctx.save();
          ctx.beginPath();
          ctx.moveTo(x, topY);
          ctx.lineTo(x, bottomY);
          ctx.lineWidth = 1;
          ctx.strokeStyle = '#999';
          ctx.stroke();
          ctx.restore();
        } else {
          $('#clones_header').text('clones in last ');
          $('#clones_count').text('0');
        }
      }
    });

    // Make the plot
    var ctx = document.getElementById('clones_plot').getContext('2d');
    new Chart(ctx, {
      data: {
        datasets: [{
          backgroundColor: 'rgba(84, 171, 106, 0.2)',
          borderColor: 'rgba(84, 171, 106, 1)',
          pointRadius: 0,
          pointHoverBorderColor: 'rgba(84, 171, 106, 0)', // transparent
          pointHoverBackgroundColor: 'rgba(84, 171, 106, 0)', // transparent
          data: [
                      ]
        }],
      },
      type: 'LineWithLine',
      options: {
        onClick: function(e) {
          window.location.href = '/16S_pipeline/stats';
        },
        elements: {
          point: {
            radius: 0,
            hitRadius: 3,
            hoverRadius: 3
          },
          line: {
            borderWidth: 2,
            tension: 0 // disables bezier curves
          }
        },
        scales: {
          xAxes: [{
            type: 'time',
            display: false
          }],
          yAxes: [{
            display: false
          }],
        },
        legend: {
          display: false
        },
        tooltips: {
          enabled: false,
          mode: 'x',
          intersect: false,
          custom: function(tooltipModel) {
            tooltipModel.opacity = 0
          },
          callbacks: {
            // Use the footer callback to display the sum of the items showing in the tooltip
            footer: function(tooltipItems, data) {
              $('#clones_header').text('clones - ' + tooltipItems[0]['label']);
              $('#clones_count').text(tooltipItems[0]['value']);
            },
          }
        },
      }
    });
  });
</script>



</body></html>
